src/rt-app.c

```diff
@@ -469,19 +463,30 @@ shutdown(int sig)
 			pthread_cond_broadcast(&opts.resources[i].res.barrier.c_obj);
 		}
 	}
+}
+
+static void __shutdown(bool force_terminate)
+{
+	int i;
+
+	if(!continue_running)
+		return;
+
+	if (force_terminate) {
+		continue_running = 0;
+		wakeup_all_threads();
+	}
 
 	/*
-	 * If we try to interrupt pthread_join() at the end of main then we
-	 * might end up joining the same thread twice which is undefined
-	 * behaviour and can lead to segfaults.
+	 * Make sure pthread_join() is done once or we hit undefined behaviour
+	 * of multiple simulataneous calls to pthread_join().
 	 *
-	 * Only join if nowhere else started the join process already.
+	 * This is a problem because __shutdown() could be called
+	 * asynchronously from some signals and from main().
 	 */
-	if (joining)
+	if (pthread_mutex_trylock(&joining_mutex))
```

Comment:
Wonder if it's OK that owner doesn't release the mutex before terminating.
I feel uncomfortable about it. :-)

Reply:
I chose not to release it because it should be held until we exit, and IMHO an implicit release by exit is the more accurate one. But FWIW, it shouldn't really matter.

I'll create a new PR with this requested change applied. Is this the only remaining issue for this to get merged? If not, can you please aggregate them all in one go so that I'll minimize my work :-)

Reply:
> an implicit release by exit

Is this documented somewhere?
Couldn't find that this is indeed the expected behaviour anywhere. :-/

Reply:
On call to exit() we terminate the application along with all of its resources. That's what I meant by implicitly releasing the mutex. The call to unlock() before exit() doesn't add anything to correctness, in contrary one could argue a potential a race between releasing the lock and calling exit() - a late ctrl+c before we just call exit() for instance?

Reply:
Guess this is true.
I don't think I have any other comment to this PR (apart from if you could please squash).
Thanks!

Reply:
I'll add a comment here anyway and I'll do the squash. Thanks for the review :)